src/DotNetWorker.Core/Context/Features/DefaultInputConversionFeature.cs

```diff
@@ -0,0 +1,137 @@
+using Microsoft.Azure.Functions.Worker.Converters;
+
+namespace Microsoft.Azure.Functions.Worker.Context.Features
+{
+    /// <summary>
+    /// Default implementation of <see cref="IInputConversionFeature"/>
+    /// </summary>
+    internal sealed class DefaultInputConversionFeature : IInputConversionFeature
+    {
+        private readonly IInputConverterProvider _inputConverterProvider;
+        private static readonly Type _inputConverterAttributeType = typeof(InputConverterAttribute);
+
+        // Users may create a POCO and specify a special converter implementation
+        // to be used using "InputConverter" attribute. We cache that mapping here.
+        // Key is assembly qualified name of POCO and Value is assembly qualified name of converter implementation.
+        private static readonly ConcurrentDictionary<string, string?> _typeToConverterCache = new();
+
+        public DefaultInputConversionFeature(IInputConverterProvider inputConverterProvider)
+        {
+            _inputConverterProvider = inputConverterProvider ?? throw new ArgumentNullException(nameof(inputConverterProvider));
+        }
+
+        /// <summary>
+        /// Executes a conversion operation with the context information provided.
+        /// </summary>
+        /// <param name="converterContext">The converter context.</param>
+        /// <returns>An instance of <see cref="ConversionResult"/> representing the result of the conversion.</returns>
+        public async ValueTask<ConversionResult> ConvertAsync(ConverterContext converterContext)
+        {
+            // Check a converter is explicitly specified via the converter context. If so, use that.
+            IInputConverter? converterFromContext = GetConverterFromContext(converterContext);
+
+            if (converterFromContext is not null)
+            {
+                var conversionResult = await ConvertAsyncUsingConverter(converterFromContext, converterContext);
+
+                if (conversionResult.IsHandled)
+                {
+                    return conversionResult;
+                }
+            }
+
+            // The first converter which can handle the conversion wins.
+            foreach (var converter in _inputConverterProvider.DefaultConverters)
+            {
+                var conversionResult = await ConvertAsyncUsingConverter(converter, converterContext);
+
+                if (conversionResult.IsHandled)
+                {
+                    return conversionResult;
+                }
+
+                // If "IsHandled" is false, we move on to the next converter and try to convert with that.
+            }
+
+            return ConversionResult.Unhandled();
+        }
+
+        private ValueTask<ConversionResult> ConvertAsyncUsingConverter(IInputConverter converter, ConverterContext context)
+        {
+            var conversionResultTask = converter.ConvertAsync(context);
+
+            if (conversionResultTask.IsCompletedSuccessfully)
```

Comment:
Do we need to check if `conversionResultTask` can be null or empty here?

Reply:
it cannot be null (value type). empty struct is same as a result which is representing an unhandled input conversion. I will revisit this part after we land a solution on the confusion related to the "unhandled" property on ConversionResult.